crypto/x509v3/v3_purp.c

```diff
@@ -760,8 +760,12 @@ int X509_check_issued(X509 *issuer, X509 *subject)
     if (X509_NAME_cmp(X509_get_subject_name(issuer),
                       X509_get_issuer_name(subject)))
         return X509_V_ERR_SUBJECT_ISSUER_MISMATCH;
+    CRYPTO_THREAD_write_lock(issuer->lock);
     x509v3_cache_extensions(issuer);
+    CRYPTO_THREAD_unlock(issuer->lock);
+    CRYPTO_THREAD_write_lock(subject->lock);
     x509v3_cache_extensions(subject);
+    CRYPTO_THREAD_unlock(subject->lock);
```

Comment:
If one looks at the other locations where `x509v3_cache_extensions()` is called, one sees that the `EXFLAG_SET` flag is used to ensure that the values are cached only once. So either the guards should be added here, too
```
    if (!(issuer->ex_flags & EXFLAG_SET)) {
        CRYPTO_THREAD_write_lock(issuer->lock);
        x509v3_cache_extensions(issuer);
        CRYPTO_THREAD_unlock(issuer->lock);
    }
    if (!(subject->ex_flags & EXFLAG_SET)) {
        CRYPTO_THREAD_write_lock(subject->lock);
        x509v3_cache_extensions(subject);
        CRYPTO_THREAD_write_lock(subject->lock);
    }
```
or even better the check  could be moved entirely into `x509v3_cache_extensions()`. See upcoming pull request...